### ISSUE TITLE
Disable default exclusion when include explicitly

### ISSUE THREAD

**Issue by blindpirate** [gradle/gradle#2986]
Originally discussed in #1348

Since there's no way to disable default exclusion explicitly, I propose some change in inclusion and exclusion mechanism:

- If user doesn't specify `include` and `exclude` explicitly, the default exclusion should be applied
- If user specifies some `include` which contains files excluded by default exclusion, these files should be "liberated", e.g. could be seen by `Copy` task and so on.

### Expected Behavior

No way to disable default exclusion inherited from Ant.

### Current Behavior

Stated above.

### Context

See #1348

**Comment by pun-ky** [comment]
in my opinion there should be at least a flag in `project.fileTree(baseDir, globalExcludes=true)` to remain existing behavior and be able to control that functionality. As a Gradle plugin developer now I am forced to have in my task such workaround:

```kotlin
    doFirst {
      DirectoryScanner.getDefaultExcludes().forEach { DirectoryScanner.removeDefaultExclude(it) }
    }
    doLast {
      DirectoryScanner.resetDefaultExcludes()
    }
```

and I am afraid that could introduce an impact in case of task concurrency, because of such static variable inside directory scanner implementation. 

about proposal described in this issue, I think that it is a breaking change when no `exclude` or `include` is defined, so that many Gradle users may observe regression in their builds.

to sum up, I am convinced that Gradle API need to be changed / have an option to control that default global exclusions working correctly in case of existing builds and new plugins that will want to use that control ability.

**Comment by Vampire** [comment]
@pun-ky Why do you think it is a breaking change? Currently the default excludes are always in-place, so the behavior wouldn't change in case neither `include` nor `exclude` is specified.

**Comment by pun-ky** [comment]
ok, got you right know... I was thinking that *not* is here " the default exclusion should be applied" :)

in other words, you'd like to be able to whitelist explicitly global exclusions (make them noneffective).
for me, more easy to use and understand will be a flag `useGlobalExclusions=false` while creating FileTree.

**Comment by pun-ky** [comment]
to workaround this, using Kotlin, I implemented easy to use extension function:

```kotlin
import org.apache.tools.ant.DirectoryScanner
import org.gradle.api.file.FileTree
import org.gradle.api.file.FileVisitDetails

fun FileTree.visitNoGlobalExclusions(visitor: (FileVisitDetails) -> Unit): FileTree {
  var reset = false
  val current = DirectoryScanner.getDefaultExcludes()

  current.forEach { DirectoryScanner.removeDefaultExclude(it) }
  return visit({ f ->
    if (!reset) {
      current.forEach { DirectoryScanner.addDefaultExclude(it) }
      reset = true
    }
    visitor(f)
  })
}
```

The reason that I am doing that is fact that PatternSet to Spec conversion including global exclusion is being performed in just after FileTree.visit call, so that just after calculating spec, we are able to recover previously set default exclusions to minimalize regression to other stuff affecting it.

**Comment by oehme** [comment]
Another option would be to completely remove the default excludes. Some variations include:

- deprecate by warning the user whenever a default exclude matches something, telling them to exclude it themselves
- providing a DSL on the `Settings` level to add defaults to all fileTrees so users don't have to repeat themselves

**Comment by big-guy** [comment]
This is probably too late for 5.0, but I wonder if we should consider deprecating this in 5.1 @oehme?  I've seen some weird things pop up in profiling that makes me think we're paying a unnecessary tax in every FileCollection because of this behavior.

**Comment by nigelgbanks** [comment]
This also affects custom tasks with output annotations, for example: 

```kotlin
@CacheableTask
open class NpmUpdate : Npm() {
    @OutputDirectory
    @PathSensitive(PathSensitivity.RELATIVE)
    val nodeModules = project.layout.projectDirectory.dir("node_modules")
}
```

When the directory "node_modules" is cached it will be missing files, some of which are required for modules to behave correctly. 

Unfortunately in this case the **DirectoryScanner** workaround does not appear to work.

**Comment by nigelgbanks** [comment]
Actually there is a work around but it's global the changes to DirectoryScanner must be done in settings.gradle.kts

```kotlin
DirectoryScanner.getDefaultExcludes().forEach { DirectoryScanner.removeDefaultExclude(it) }
```

**Comment by nigelgbanks** [comment]
Actually disregard my work around for settings.gradle.kts it doesn't work consistently, perhaps the timing between when BuildSessionScopeServices.java is instantiated vs when settings.gradle.kts could explain the inconsistency.

**Comment by wolfs** [comment]
For 6.7, there is now a little bit of improvement regarding default excludes: Default excludes can be changed in `settings.gradle` and they will take effect immediately and consistently during a build: https://github.com/gradle/gradle/pull/13694

The change will be in Gradle 6.7.

**Comment by oehme** [comment]
@wolfs I just stumbled over this (node_modules caching use case) and updating the default excludes works, but doesn't change the cache key, so the old, broken output is still reused. I think the excludes should factor into cache keys.

**Comment by jdaugherty** [comment]
FYI: Later versions of gradle remove the ability to change the default excludes at build time.  They must be done in settings.gradle.  This means you have to opt out of the excludes everywhere and then add them back on tasks that need them.
